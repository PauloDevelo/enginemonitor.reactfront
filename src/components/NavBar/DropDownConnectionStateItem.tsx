import React, { useState, useEffect } from 'react';
import { DropdownItem } from 'reactstrap';

import { faPlug, faBan } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import ConnectionStateMessage from './ConnectionStateMessage';
import NbActionPending from './NbActionPending';

import onlineManager from '../../services/OnlineManager';
import syncService from '../../services/SyncService';
import actionManager from '../../services/ActionManager';

type ConnectionState = {
    isOnline: boolean,
    isSynced: boolean
};

// eslint-disable-next-line no-unused-vars
const ConnectionStateIcon = ({ isOnline, isSynced }:ConnectionState) => {
  if (isOnline) {
    return <FontAwesomeIcon icon={faPlug} />;
  }

  return <FontAwesomeIcon icon={faBan} />;
};

const DropDownConnectionStateItem = () => {
  const [isOnline, onIsOnlineChanged] = useState(false);
  const [isSync, onIsSynced] = useState(false);

  const onActionManagerChanged = (nbAction: number): void => {
    onIsSynced(nbAction === 0);
  };

  useEffect(() => {
    onlineManager.registerIsOnlineListener(onIsOnlineChanged);
    onlineManager.isOnline().then((online) => onIsOnlineChanged(online));

    actionManager.registerOnActionManagerChanged(onActionManagerChanged);
    onActionManagerChanged(actionManager.countAction());

    return () => {
      onlineManager.unregisterIsOnlineListener(onIsOnlineChanged);
      actionManager.unregisterOnActionManagerChanged(onActionManagerChanged);
    };
  }, []);

  return (
    <DropdownItem disabled={isOnline === false || isSync === true} onClick={() => syncService.tryToRun()}>
      <ConnectionStateIcon isOnline={isOnline} isSynced={isSync} />
      {' '}
      <ConnectionStateMessage isOnline={isOnline} isSynced={isSync} />
      (
      <NbActionPending />
      )
    </DropdownItem>
  );
};

export default DropDownConnectionStateItem;
