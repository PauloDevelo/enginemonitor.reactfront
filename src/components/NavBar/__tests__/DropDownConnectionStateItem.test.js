import React from 'react';
import { mount } from 'enzyme';

// eslint-disable-next-line no-unused-vars
import localforage from 'localforage';
import actionManager from '../../../services/ActionManager';
import syncService from '../../../services/SyncService';
import onlineManager from '../../../services/OnlineManager';

import DropDownConnectionStateItem from '../DropDownConnectionStateItem';

import ignoredMessages from '../../../testHelpers/MockConsole';
import updateWrapper from '../../../testHelpers/EnzymeHelper';

jest.mock('localforage');
jest.mock('../../../services/ActionManager');
jest.mock('../../../services/SyncService');
jest.mock('../../../services/OnlineManager');

describe('DropDownConnectionStateItem', () => {
  beforeAll(() => {
    ignoredMessages.length = 0;
    ignoredMessages.push('Could not find required `intl` object.');
    ignoredMessages.push('a test was not wrapped in act');
  });

  beforeEach(() => {
    jest.spyOn(onlineManager, 'registerIsOnlineListener');
    jest.spyOn(onlineManager, 'unregisterIsOnlineListener');

    jest.spyOn(actionManager, 'registerOnActionManagerChanged');
    jest.spyOn(actionManager, 'unregisterOnActionManagerChanged');
  });

  afterEach(async () => {
    onlineManager.registerIsOnlineListener.mockRestore();
    onlineManager.unregisterIsOnlineListener.mockRestore();
    onlineManager.isOnline.mockRestore();
    syncService.synchronize.mockRestore();

    actionManager.registerOnActionManagerChanged.mockRestore();
    actionManager.unregisterOnActionManagerChanged.mockRestore();
    actionManager.countAction.mockRestore();
  });

  const states = [
    { isOnline: true, nbAction: 0 },
    { isOnline: true, nbAction: 3 },
    { isOnline: false, nbAction: 0 },
    { isOnline: false, nbAction: 5 },
  ];
  describe.each(states)('Render', ({ isOnline, nbAction }) => {
    it(`Should render when the app is online ${isOnline} and nb action is ${nbAction} as expected `, async (done) => {
      // Arrange
      jest.spyOn(onlineManager, 'isOnline').mockImplementation(async () => Promise.resolve(isOnline));
      jest.spyOn(syncService, 'synchronize');
      jest.spyOn(actionManager, 'countAction').mockImplementation(async () => Promise.resolve(nbAction));

      // Act
      const dropDownConnectionStateItemWrapper = mount(<DropDownConnectionStateItem />);
      await updateWrapper(dropDownConnectionStateItemWrapper);

      // Assert
      expect(dropDownConnectionStateItemWrapper).toMatchSnapshot();
      expect(onlineManager.registerIsOnlineListener).toBeCalledTimes(1);
      expect(actionManager.registerOnActionManagerChanged).toBeCalledTimes(2);

      dropDownConnectionStateItemWrapper.unmount();
      expect(onlineManager.unregisterIsOnlineListener).toBeCalledTimes(1);
      expect(actionManager.unregisterOnActionManagerChanged).toBeCalledTimes(2);
      done();
    });
  });
});
