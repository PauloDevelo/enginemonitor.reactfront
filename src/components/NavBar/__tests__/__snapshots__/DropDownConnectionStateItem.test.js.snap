// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownConnectionStateItem Render Should render when the app is online false and nb action is 0 as expected  1`] = `
<DropDownConnectionStateItem>
  <DropdownItem
    disabled={true}
    onClick={[Function]}
    tag="button"
    toggle={true}
  >
    <button
      className="disabled dropdown-item"
      disabled={true}
      onClick={[Function]}
      tabIndex="-1"
      type="button"
    >
      <ConnectionStateIcon
        isOnline={false}
        isSynced={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f05e",
                "M256 8C119.034 8 8 119.033 8 256s111.034 248 248 248 248-111.034 248-248S392.967 8 256 8zm130.108 117.892c65.448 65.448 70 165.481 20.677 235.637L150.47 105.216c70.204-49.356 170.226-44.735 235.638 20.676zM125.892 386.108c-65.448-65.448-70-165.481-20.677-235.637L361.53 406.784c-70.203 49.356-170.226 44.736-235.638-20.676z",
              ],
              "iconName": "ban",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-ban fa-w-16 "
            data-icon="ban"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119.034 8 8 119.033 8 256s111.034 248 248 248 248-111.034 248-248S392.967 8 256 8zm130.108 117.892c65.448 65.448 70 165.481 20.677 235.637L150.47 105.216c70.204-49.356 170.226-44.735 235.638 20.676zM125.892 386.108c-65.448-65.448-70-165.481-20.677-235.637L361.53 406.784c-70.203 49.356-170.226 44.736-235.638-20.676z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </FontAwesomeIcon>
      </ConnectionStateIcon>
       
      <Memo(ConnectionStateMessage)
        isOnline={false}
        isSynced={true}
      >
        <FormattedMessage
          defaultMessage="Disconnected"
          id="notonline"
          values={Object {}}
        >
          Disconnected
        </FormattedMessage>
      </Memo(ConnectionStateMessage)>
      (
      <NbActionPending>
        <FormattedMessage
          defaultMessage="{nbAction, plural, =0 {synced} one {# action to sync} other {# actions to sync}}"
          description="Nb action pending"
          id="nbAction"
          values={
            Object {
              "nbAction": 0,
            }
          }
        >
          synced
        </FormattedMessage>
      </NbActionPending>
      )
    </button>
  </DropdownItem>
</DropDownConnectionStateItem>
`;

exports[`DropDownConnectionStateItem Render Should render when the app is online false and nb action is 5 as expected  1`] = `
<DropDownConnectionStateItem>
  <DropdownItem
    disabled={true}
    onClick={[Function]}
    tag="button"
    toggle={true}
  >
    <button
      className="disabled dropdown-item"
      disabled={true}
      onClick={[Function]}
      tabIndex="-1"
      type="button"
    >
      <ConnectionStateIcon
        isOnline={false}
        isSynced={false}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f05e",
                "M256 8C119.034 8 8 119.033 8 256s111.034 248 248 248 248-111.034 248-248S392.967 8 256 8zm130.108 117.892c65.448 65.448 70 165.481 20.677 235.637L150.47 105.216c70.204-49.356 170.226-44.735 235.638 20.676zM125.892 386.108c-65.448-65.448-70-165.481-20.677-235.637L361.53 406.784c-70.203 49.356-170.226 44.736-235.638-20.676z",
              ],
              "iconName": "ban",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-ban fa-w-16 "
            data-icon="ban"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119.034 8 8 119.033 8 256s111.034 248 248 248 248-111.034 248-248S392.967 8 256 8zm130.108 117.892c65.448 65.448 70 165.481 20.677 235.637L150.47 105.216c70.204-49.356 170.226-44.735 235.638 20.676zM125.892 386.108c-65.448-65.448-70-165.481-20.677-235.637L361.53 406.784c-70.203 49.356-170.226 44.736-235.638-20.676z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </FontAwesomeIcon>
      </ConnectionStateIcon>
       
      <Memo(ConnectionStateMessage)
        isOnline={false}
        isSynced={false}
      >
        <FormattedMessage
          defaultMessage="Disconnected"
          id="notonline"
          values={Object {}}
        >
          Disconnected
        </FormattedMessage>
      </Memo(ConnectionStateMessage)>
      (
      <NbActionPending>
        <FormattedMessage
          defaultMessage="{nbAction, plural, =0 {synced} one {# action to sync} other {# actions to sync}}"
          description="Nb action pending"
          id="nbAction"
          values={
            Object {
              "nbAction": 5,
            }
          }
        >
          5 actions to sync
        </FormattedMessage>
      </NbActionPending>
      )
    </button>
  </DropdownItem>
</DropDownConnectionStateItem>
`;

exports[`DropDownConnectionStateItem Render Should render when the app is online true and nb action is 0 as expected  1`] = `
<DropDownConnectionStateItem>
  <DropdownItem
    disabled={true}
    onClick={[Function]}
    tag="button"
    toggle={true}
  >
    <button
      className="disabled dropdown-item"
      disabled={true}
      onClick={[Function]}
      tabIndex="-1"
      type="button"
    >
      <ConnectionStateIcon
        isOnline={true}
        isSynced={true}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                384,
                512,
                Array [],
                "f1e6",
                "M320,32a32,32,0,0,0-64,0v96h64Zm48,128H16A16,16,0,0,0,0,176v32a16,16,0,0,0,16,16H32v32A160.07,160.07,0,0,0,160,412.8V512h64V412.8A160.07,160.07,0,0,0,352,256V224h16a16,16,0,0,0,16-16V176A16,16,0,0,0,368,160ZM128,32a32,32,0,0,0-64,0v96h64Z",
              ],
              "iconName": "plug",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-plug fa-w-12 "
            data-icon="plug"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 384 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320,32a32,32,0,0,0-64,0v96h64Zm48,128H16A16,16,0,0,0,0,176v32a16,16,0,0,0,16,16H32v32A160.07,160.07,0,0,0,160,412.8V512h64V412.8A160.07,160.07,0,0,0,352,256V224h16a16,16,0,0,0,16-16V176A16,16,0,0,0,368,160ZM128,32a32,32,0,0,0-64,0v96h64Z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </FontAwesomeIcon>
      </ConnectionStateIcon>
       
      <Memo(ConnectionStateMessage)
        isOnline={true}
        isSynced={true}
      >
        <FormattedMessage
          defaultMessage="Online"
          id="online"
          values={Object {}}
        >
          Online
        </FormattedMessage>
      </Memo(ConnectionStateMessage)>
      (
      <NbActionPending>
        <FormattedMessage
          defaultMessage="{nbAction, plural, =0 {synced} one {# action to sync} other {# actions to sync}}"
          description="Nb action pending"
          id="nbAction"
          values={
            Object {
              "nbAction": 0,
            }
          }
        >
          synced
        </FormattedMessage>
      </NbActionPending>
      )
    </button>
  </DropdownItem>
</DropDownConnectionStateItem>
`;

exports[`DropDownConnectionStateItem Render Should render when the app is online true and nb action is 3 as expected  1`] = `
<DropDownConnectionStateItem>
  <DropdownItem
    disabled={false}
    onClick={[Function]}
    tag="button"
    toggle={true}
  >
    <button
      className="dropdown-item"
      disabled={false}
      onClick={[Function]}
      role="menuitem"
      tabIndex="0"
      type="button"
    >
      <ConnectionStateIcon
        isOnline={true}
        isSynced={false}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                384,
                512,
                Array [],
                "f1e6",
                "M320,32a32,32,0,0,0-64,0v96h64Zm48,128H16A16,16,0,0,0,0,176v32a16,16,0,0,0,16,16H32v32A160.07,160.07,0,0,0,160,412.8V512h64V412.8A160.07,160.07,0,0,0,352,256V224h16a16,16,0,0,0,16-16V176A16,16,0,0,0,368,160ZM128,32a32,32,0,0,0-64,0v96h64Z",
              ],
              "iconName": "plug",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-plug fa-w-12 "
            data-icon="plug"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 384 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320,32a32,32,0,0,0-64,0v96h64Zm48,128H16A16,16,0,0,0,0,176v32a16,16,0,0,0,16,16H32v32A160.07,160.07,0,0,0,160,412.8V512h64V412.8A160.07,160.07,0,0,0,352,256V224h16a16,16,0,0,0,16-16V176A16,16,0,0,0,368,160ZM128,32a32,32,0,0,0-64,0v96h64Z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </FontAwesomeIcon>
      </ConnectionStateIcon>
       
      <Memo(ConnectionStateMessage)
        isOnline={true}
        isSynced={false}
      >
        <FormattedMessage
          defaultMessage="Synchronize"
          id="syncAction"
          values={Object {}}
        >
          Synchronize
        </FormattedMessage>
      </Memo(ConnectionStateMessage)>
      (
      <NbActionPending>
        <FormattedMessage
          defaultMessage="{nbAction, plural, =0 {synced} one {# action to sync} other {# actions to sync}}"
          description="Nb action pending"
          id="nbAction"
          values={
            Object {
              "nbAction": 3,
            }
          }
        >
          3 actions to sync
        </FormattedMessage>
      </NbActionPending>
      )
    </button>
  </DropdownItem>
</DropDownConnectionStateItem>
`;
